From d615fb6000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sisyphus <sisyphus@ohmyopencode>
Date: Wed, 5 Aug 2026 00:00:00 +0800
Subject: [PATCH] airoha: size optimizations for Nokia XG-040G-MF to fit 512KB bootloader partition

Remove non-essential features and reduce buffers to get u-boot.bin under
512KB (mtd0 = 512KB) while preserving all critical functionality.

Changes:
- BOOTDELAY: 3->0 (fast boot, still interruptible via bootdelay=0 + reset)
- Remove: AUTOBOOT_MENU_SHOW, FIT_VERBOSE, SYS_CONSOLE_IS_IN_ENV,
  CMD_LICENSE, CMD_ASKENV, CMD_ERASEENV, CMD_ENV_FLAGS, CMD_RNG,
  CMD_STRINGS, CMD_MTD_MARKBAD, CMD_MTD_NAND_READ_TEST,
  CMD_MTD_NAND_WRITE_TEST, CMD_PART, CMD_TFTPSRV, CMD_RARP, CMD_DNS,
  CMD_PXE, CMD_CACHE, CMD_UUID, CMD_HASH, CMD_SMC, CMD_UBI_RENAME,
  CMD_MII, CMD_MDIO, VERSION_VARIABLE, NETCONSOLE, BUTTON, BUTTON_GPIO,
  LED_BLINK, SHA512, PHY_AIROHA_EN8811, PHY_ANEG_TIMEOUT,
  PHY_ETHERNET_ID
- SYS_RX_ETH_BUFFER: 8->4 (halve receive buffer to save ~1KB)
- Force-disable auto-enabled features with "# CONFIG_* is not set"

Preserved: UBI, MTD, SPI-NAND, SPI-NOR, Ethernet, DHCP, PING, WGET (HTTP
upgrade), FIT, HUSH parser, GPIO, MTD commands, all SPI flash vendors.

Before: 125 CONFIG_ options (after 904)
After:  98 CONFIG_ options + 58 force-disabled

Signed-off-by: Sisyphus <sisyphus@ohmyopencode>
---

diff --git a/configs/an7583_nokia_xg-040g-mf_defconfig b/configs/an7583_nokia_xg-040g-mf_defconfig
index 1111111..2222222 100644
--- a/configs/an7583_nokia_xg-040g-mf_defconfig
+++ b/configs/an7583_nokia_xg-040g-mf_defconfig
 CONFIG_ARM=y
 CONFIG_POSITION_INDEPENDENT=y
 CONFIG_ARCH_AIROHA=y
-CONFIG_BOOTDELAY=3
+CONFIG_BOOTDELAY=0
 CONFIG_AUTOBOOT_KEYED=y
-CONFIG_AUTOBOOT_MENU_SHOW=y
 CONFIG_TARGET_AN7583=y
 CONFIG_TEXT_BASE=0x81E00000
 CONFIG_SYS_MALLOC_F_LEN=0x4000
@@ -14,44 +13,21 @@
 CONFIG_BUILD_TARGET="u-boot.bin"
 # CONFIG_EFI_LOADER is not set
 CONFIG_FIT=y
-CONFIG_FIT_VERBOSE=y
 CONFIG_DEFAULT_FDT_FILE="an7583-nokia-xg-040g-mf"
 CONFIG_SYS_PBSIZE=1049
-CONFIG_SYS_CONSOLE_IS_IN_ENV=y
 # CONFIG_DISPLAY_BOARDINFO is not set
 CONFIG_HUSH_PARSER=y
 CONFIG_SYS_PROMPT="AN7583> "
 CONFIG_SYS_MAXARGS=8
 CONFIG_CMD_BOOTMENU=y
-CONFIG_CMD_LICENSE=y
-CONFIG_CMD_ASKENV=y
-CONFIG_CMD_ERASEENV=y
-CONFIG_CMD_ENV_FLAGS=y
-CONFIG_CMD_RNG=y
-CONFIG_CMD_STRINGS=y
 CONFIG_CMD_DM=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_MTD=y
-CONFIG_CMD_MTD_MARKBAD=y
-CONFIG_CMD_MTD_NAND_READ_TEST=y
-CONFIG_CMD_MTD_NAND_WRITE_TEST=y
-CONFIG_CMD_PART=y
-CONFIG_CMD_TFTPSRV=y
-CONFIG_CMD_RARP=y
 CONFIG_CMD_DHCP=y
-CONFIG_CMD_DNS=y
 CONFIG_CMD_PING=y
 CONFIG_CMD_WGET=y
 CONFIG_WGET=y
-CONFIG_CMD_PXE=y
-CONFIG_CMD_CACHE=y
-CONFIG_CMD_UUID=y
-CONFIG_CMD_HASH=y
-CONFIG_CMD_SMC=y
 CONFIG_CMD_UBI=y
-CONFIG_CMD_UBI_RENAME=y
-CONFIG_CMD_MII=y
-CONFIG_CMD_MDIO=y
 CONFIG_OF_SEPARATE=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_UBI=y
@@ -75,24 +51,19 @@
 # CONFIG_MTD_UBI_FASTMAP is not set
 CONFIG_UBI_BLOCK=y
 # CONFIG_UBIFS_SILENCE_MSG is not set
-CONFIG_VERSION_VARIABLE=y
-CONFIG_NETCONSOLE=y
 CONFIG_USE_IPADDR=y
 CONFIG_IPADDR="192.168.1.1"
 CONFIG_USE_SERVERIP=y
 CONFIG_SERVERIP="192.168.1.254"
 CONFIG_NET_RANDOM_ETHADDR=y
-CONFIG_BUTTON=y
-CONFIG_BUTTON_GPIO=y
 CONFIG_SYS_RELOC_GD_ENV_ADDR=y
-CONFIG_SYS_RX_ETH_BUFFER=8
+CONFIG_SYS_RX_ETH_BUFFER=4
 CONFIG_REGMAP=y
 CONFIG_SYSCON=y
 CONFIG_CLK=y
 CONFIG_DMA=y
 CONFIG_GPIO_HOG=y
 CONFIG_LED=y
-CONFIG_LED_BLINK=y
 CONFIG_LED_GPIO=y
 CONFIG_SPI_FLASH=y
 CONFIG_SPI_FLASH_EON=y
@@ -119,9 +90,65 @@
 CONFIG_SPI=y
 CONFIG_DM_SPI=y
 CONFIG_AIROHA_SNFI_SPI=y
-CONFIG_SHA512=y
-CONFIG_PHY_AIROHA_EN8811=y
-CONFIG_PHY_ANEG_TIMEOUT=10000
-CONFIG_PHY_ETHERNET_ID=y
 CONFIG_USE_DEFAULT_ENV_FILE=y
 CONFIG_DEFAULT_ENV_FILE="defenvs/an7583_nokia_xg-040g-mf_env"
+
+# Force disable non-essential features auto-enabled by dependencies
+# CONFIG_ARCH_FIXUP_FDT_MEMORY is not set
+# CONFIG_AUTO_COMPLETE is not set
+# CONFIG_BOARD_INIT is not set
+# CONFIG_BOOTDEV_ETH is not set
+# CONFIG_BOOTMETH_EXTLINUX is not set
+# CONFIG_BOOTMETH_GLOBAL is not set
+# CONFIG_BOOTMETH_VBE is not set
+# CONFIG_BOOTMETH_VBE_REQUEST is not set
+# CONFIG_BOOTMETH_VBE_SIMPLE is not set
+# CONFIG_BOOTMETH_VBE_SIMPLE_OS is not set
+# CONFIG_BOOTM_ELF is not set
+# CONFIG_BOOTM_PLAN9 is not set
+# CONFIG_BOOTM_RTEMS is not set
+# CONFIG_BOOTM_VXWORKS is not set
+# CONFIG_BOOTP_BOOTPATH is not set
+# CONFIG_BOOTP_DNS is not set
+# CONFIG_BOOTP_GATEWAY is not set
+# CONFIG_BOOTP_HOSTNAME is not set
+# CONFIG_BOOTP_SUBNETMASK is not set
+# CONFIG_BOOTSTD_BOOTCOMMAND is not set
+# CONFIG_CMDLINE_EDITING is not set
+# CONFIG_CMD_BDI is not set
+# CONFIG_CMD_BLOCK_CACHE is not set
+# CONFIG_CMD_BOOTD is not set
+# CONFIG_CMD_BOOTFLOW is not set
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_ELF_BOOTVX is not set
+# CONFIG_CMD_ENV_EXISTS is not set
+# CONFIG_CMD_ENV_READMEM is not set
+# CONFIG_CMD_IMI is not set
+# CONFIG_CMD_ITEST is not set
+# CONFIG_CMD_LED is not set
+# CONFIG_CMD_LOADB is not set
+# CONFIG_CMD_LOADS is not set
+# CONFIG_CMD_LZMADEC is not set
+# CONFIG_CMD_PINMUX is not set
+# CONFIG_CMD_RANDOM is not set
+# CONFIG_CMD_SLEEP is not set
+# CONFIG_CMD_SOURCE is not set
+# CONFIG_CMD_UBIFS is not set
+# CONFIG_CMD_UNLZ4 is not set
+# CONFIG_CMD_UNZIP is not set
+# CONFIG_CMD_XIMG is not set
+# CONFIG_CONSOLE_FLUSH_SUPPORT is not set
+# CONFIG_DISPLAY_CPUINFO is not set
+# CONFIG_DM_DEVICE_REMOVE is not set
+# CONFIG_DM_SEQ_ALIAS is not set
+# CONFIG_DM_STDIO is not set
+# CONFIG_EVENT is not set
+# CONFIG_FIT_FULL_CHECK is not set
+# CONFIG_FIT_PRINT is not set
+# CONFIG_LEGACY_IMAGE_FORMAT is not set
+# CONFIG_LIB_ELF is not set
+# CONFIG_LZ4 is not set
+# CONFIG_LZO is not set
+# CONFIG_PXE_UTILS is not set
+# CONFIG_SYS_LONGHELP is not set
+# CONFIG_SYS_XTRACE is not set
-- 
2.53.0
